<a href="https://colab.research.google.com/github/erfano10/Modul_2KKA_Erfano_Alfarrezal_Raditya_XI_RPL_2/blob/main/Modul_2__Erfano_Alfarrrezal_Raditya.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
harga = np.array([5000, 7000, 3000, 12000, 4500])
print('Rata-rata harga:', harga.mean())
print('Harga tertinggi:', harga.max())
print('Harga setelah diskon 10%:', harga * 0.9)

Rata-rata harga: 6300.0
Harga tertinggi: 12000
Harga setelah diskon 10%: [ 4500.  6300.  2700. 10800.  4050.]


Tugas Analisis 1
Jika mengalikan list biasa Python dengan float ([5000, 7000, 3000] * 0.9), program akan menghasilkan TypeError karena operator perkalian pada list Python tidak mendukung operasi matematika per elemen—pada list, perkalian hanya berfungsi mengulang isi (duplikasi) dan wajib menggunakan bilangan bulat (integer). Sebaliknya, NumPy Array dirancang khusus untuk komputasi numerik dengan fitur broadcasting, sehingga perkalian 0.9 otomatis diteruskan ke setiap elemen di dalam array untuk menghasilkan nilai terdiskon [4500. 6300. 2700. 10800. 4050.].

In [ ]:
import pandas as pd
data_kantin = {
'menu': ['Nasi Goreng', 'Es Teh', 'Mie Ayam', 'Es Teh', None],
'harga': [12000, 4000, 10000, 4000, 8000],
'terjual': [23, 40, None, 35, 18]
}
df = pd.DataFrame(data_kantin)
print(df)

          menu  harga  terjual
0  Nasi Goreng  12000     23.0
1       Es Teh   4000     40.0
2     Mie Ayam  10000      NaN
3       Es Teh   4000     35.0
4         None   8000     18.0


Latihan Analisis 2
Kolom yang memiliki data kosong adalah terjual (pada baris ke-2 bertuliskan NaN) dan menu (pada baris ke-4 bertuliskan None). Risikonya, hasil analisis statistik menjadi tidak akurat.

In [ ]:
df = pd.read_csv('data_kantin.csv')
print(df.head()) # 5 baris pertama
print(df.info()) # tipe data & jumlah non-null tiap kolom
print(df.describe()) # statistik ringkas kolom numerik
print(df.shape) # jumlah (baris, kolom)

      tanggal         menu kategori  harga  terjual
0  2026-08-20   Teh Hangat  Minuman   3000     11.0
1  2026-08-31  Nasi Goreng  Makanan  12000     29.0
2  2026-08-21  Nasi Goreng  Makanan  12000     28.0
3  2026-08-28   Teh Hangat  Minuman   3000     12.0
4  2026-08-21     Es Jeruk  Minuman   5000     19.0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110 entries, 0 to 109
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tanggal   110 non-null    object 
 1   menu      107 non-null    object 
 2   kategori  110 non-null    object 
 3   harga     110 non-null    int64  
 4   terjual   104 non-null    float64
dtypes: float64(1), int64(1), object(3)
memory usage: 4.4+ KB
None
              harga     terjual
count    110.000000  104.000000
mean    6336.363636   22.923077
std     3520.001232   10.831178
min     2000.000000    8.000000
25%     3000.000000   13.000000
50%     5000.000000   20.000000
75%    10000.000000 

Latihan Analisis 3
Dari data total 110 baris, kolom yang jumlah non-null-nya kurang dari 110 adalah:

menu (hanya ada 107 non-null)

dan terjual (hanya ada 104 non-null)

Artinya: Kolom-kolom tersebut memiliki data yang hilang atau kosong (missing values). Terdapat 3 baris yang tidak memiliki nama menu dan 6 baris yang tidak memiliki angka jumlah penjualan.

In [ ]:
print(df.isnull().sum()) # jumlah data kosong tiap kolom
df['terjual'] = df['terjual'].fillna(0)
df = df.dropna(subset=['menu']) # hapus baris jika kolom menu kosong

menu       1
harga      0
terjual    1
dtype: int64


Latihan Analisis 4
terjual diisi fillna(0) karena berupa angka numerik, sehingga data kosong aman diasumsikan sebagai 0 unit terjual tanpa membuang data lainnya. menu dihapus dropna() karena merupakan identitas produk—catatan transaksi menjadi tidak bermakna jika nama menunya tidak diketahui.

In [ ]:
print(df.duplicated().sum()) # jumlah baris duplikat
df = df.drop_duplicates()
df['harga'] = df['harga'].astype(int) # memastikan tipe data harga adalah integer
print(df.dtypes)

0
menu        object
harga        int64
terjual    float64
dtype: object


Tugas Analisis 5
Berdasarkan output 0 duplikat yang ditemukan, jumlah baris dataset sekarang tetap sama (tidak berkurang sama sekali) dibanding sebelumnya.

Memastikan tipe data sudah benar sangat penting agar operasi analisis berjalan akurat dan tidak error—misalnya, jika harga bertipe teks (string), kita tidak akan bisa melakukan perhitungan matematika seperti penjumlahan total omzet atau pencarian rata-rata.

In [ ]:
laris = df[df['terjual'] > 20] # filtering
urut = df.sort_values(by='terjual', ascending=False) # sorting
df['total_pendapatan'] = df['harga'] * df['terjual'] # kolom turunan
ringkasan = df.groupby('menu')['total_pendapatan'].sum() # agregasi
print(ringkasan)

menu
Es Teh         300000.0
Mie Ayam            0.0
Nasi Goreng    276000.0
Name: total_pendapatan, dtype: float64


Tugas Analisis 6
Es Teh dengan total pendapatan sebesar 300.000,0
Informasi ini membantu pengelola kantin mengambil keputusan strategis, seperti menjaga ketersediaan stok bahan baku Es Teh agar tidak kehabisan karena merupakan sumber pendapatan utama, mengevaluasi atau mengganti menu Mie Ayam yang tidak menghasilkan pendapatan sama sekali, serta membuat strategi pemasaran baru seperti paket bundling (Nasi Goreng + Es Teh) untuk memaksimalkan omzet harian.

In [38]:
import pandas as pd
import numpy as np
from google.colab import files

# ==============================================================================
# 1. BACA DATASET AWAL & INSPEKSI
# ==============================================================================
df = pd.read_csv('dataset_penjualan_kantin.csv')

print("--- 5 Baris Pertama ---")
print(df.head(5))

print("\n--- Info Dataset ---")
df.info()

print("\n--- Jumlah Missing Value ---")
print(df.isnull().sum())

print(f"\n--- Jumlah Duplikat: {df.duplicated().sum()} ---")


# ==============================================================================
# 2. FUNGSI KHUSUS PEMBERSIH TANGGAL
# ==============================================================================
def bersihkan_tanggal(val):
    if pd.isna(val):
        return val
    val_str = str(val).strip()

    bulan_dict = {
        'januari': '01', 'februari': '02', 'maret': '03', 'april': '04',
        'mei': '05', 'juni': '06', 'juli': '07', 'agustus': '08',
        'september': '09', 'oktober': '10', 'november': '11', 'desember': '12'
    }

    # Format "14 Agustus 2026"
    for bln_nama, bln_angka in bulan_dict.items():
        if bln_nama in val_str.lower():
            bagian = val_str.split()
            if len(bagian) == 3:
                tgl = bagian[0].zfill(2)
                thn = bagian[2]
                return f"{thn}-{bln_angka}-{tgl}"

    # Format "13/08/2026"
    if '/' in val_str:
        bagian = val_str.split('/')
        if len(bagian) == 3:
            tgl = bagian[0].zfill(2)
            bln = bagian[1].zfill(2)
            thn = bagian[2]
            return f"{thn}-{bln}-{tgl}"

    return val_str


# ==============================================================================
# 3. PROSES PEMBERSIHAN DATASET
# ==============================================================================
df_clean = df.copy()

# A. Hapus Duplikat
df_clean = df_clean.drop_duplicates().reset_index(drop=True)

# B. Standarisasi Teks Kategori
df_clean['kategori'] = df_clean['kategori'].str.capitalize()

# C. Bersihkan Format Harga Satuan
df_clean['harga_satuan'] = df_clean['harga_satuan'].astype(str).str.replace('Rp', '', regex=False)
df_clean['harga_satuan'] = df_clean['harga_satuan'].str.replace('.', '', regex=False).str.strip()
df_clean['harga_satuan'] = pd.to_numeric(df_clean['harga_satuan'], errors='coerce')

harga_produk_map = {
    'Nasi Goreng': 12000, 'Mie Ayam': 10000, 'Bakso': 11000,
    'Soto Ayam': 12000, 'Gorengan': 2000, 'Roti Bakar': 7000,
    'Kerupuk': 2000, 'Es Teh': 3000, 'Es Jeruk': 4000,
    'Jus Alpukat': 8000, 'Kopi': 5000, 'Air Mineral': 3000
}
df_clean['harga_satuan'] = df_clean['harga_satuan'].fillna(df_clean['nama_produk'].map(harga_produk_map))

# D. Bersihkan Jumlah Terjual
df_clean['jumlah_terjual'] = df_clean['jumlah_terjual'].astype(str).str.replace('pcs', '', case=False, regex=False).str.strip()
df_clean['jumlah_terjual'] = pd.to_numeric(df_clean['jumlah_terjual'], errors='coerce')
df_clean['jumlah_terjual'] = df_clean['jumlah_terjual'].fillna(df_clean['jumlah_terjual'].median())

# E. Bersihkan Tanggal & Kasir
df_clean['tanggal'] = df_clean['tanggal'].ffill().bfill()
df_clean['tanggal'] = df_clean['tanggal'].apply(bersihkan_tanggal)
df_clean['nama_kasir'] = df_clean['nama_kasir'].fillna(df_clean['nama_kasir'].mode()[0])

# F. Contoh Menggunakan NumPy: Mengubah tipe data ke np.int64 & Menghitung Total
df_clean['jumlah_terjual'] = df_clean['jumlah_terjual'].astype(np.int64)
df_clean['harga_satuan'] = df_clean['harga_satuan'].astype(np.int64)
df_clean['total_penjualan'] = np.multiply(df_clean['jumlah_terjual'], df_clean['harga_satuan'])


# ==============================================================================
# 4. RINGKASAN OUTPUT TERAKHIR
# ==============================================================================
ringkasan_kategori = df_clean.groupby('kategori')[['jumlah_terjual', 'total_penjualan']].sum().reset_index()

print("\n--- Ringkasan Penjualan per Kategori ---")
print(ringkasan_kategori)


# ==============================================================================
# 5. SIMPAN DATASET BERSIH
# ==============================================================================
df_clean.to_csv('dataset_bersih.csv', index=False)
print("\nDataset bersih berhasil disimpan ke 'dataset_bersih.csv'")

try:
    files.download('dataset_bersih.csv')
except Exception:
    pass

--- 5 Baris Pertama ---
  id_transaksi     tanggal  nama_produk kategori jumlah_terjual harga_satuan  \
0      TRX0042  2026-08-11   Roti Bakar  Makanan              9         7000   
1      TRX0005  2026-08-03     Gorengan  makanan              2         2000   
2      TRX0011  2026-08-04  Jus Alpukat  Minuman              4          NaN   
3      TRX0035  2026-08-10     Mie Ayam  Makanan            NaN        10000   
4      TRX0007  2026-08-03      Kerupuk    Snack             10      Rp2.000   

  nama_kasir metode_pembayaran  
0   Pak Agus             Tunai  
1     Bu Sri              QRIS  
2    Bu Wati             Tunai  
3   Pak Joko          Transfer  
4    Bu Wati          Transfer  

--- Info Dataset ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69 entries, 0 to 68
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   id_transaksi       69 non-null     object
 1   tanggal            69 non-n

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Nama Anggota
Kelompok

1. M Haidar K.A
2. Erfano Alfarrezal R

Dataset yang Dipilih □ Data Penjualan Kantin

Pertanyaan Analisis
Awal

1. Produk apa yang paling laris berdasarkan total kuantitas penjualan?

2. Berapa total pendapatan kantin per kategori produk?

3. Siapa metode pembayaran yang paling sering digunakan oleh pembeli?    

Dugaan Masalah
Kualitas Data

1. □ Missing value pada kolom: jumlah_terjual,Harga_Satuan dan nama_kasir
2. □ Duplikat data : Ditemukan 4 baris data duplikat.
3. □ Tipe data
 & tidak sesuai pada kolom: tanggal: Format tanggal bercampur (contoh: 2026-08-11, 13/08/2026, dan 3 Agustus 2026).

kategori: Format huruf besar/kecil tidak konsisten (contoh: Makanan, makanan, MAKANAN).

harga_satuan: Berisi teks currency/simbol (contoh: Rp2.000 vs 7000).

jumlah_terjual: Berisi string teks (contoh: 12 pcs) dan nilai anomali/outlier (500).

Rencana Teknik
Pembersihan

1. Menghapus baris duplikat menggunakan drop_duplicates().

2. Menyeragamkan format teks pada kolom kategori menggunakan .str.title().

3. Membersihkan karakter string (seperti Rp, ., pcs) pada kolom harga_satuan dan jumlah_terjual lalu mengubah tipe datanya menjadi numerik (int/float).

4. Mengisi missing value pada harga_satuan sesuai dengan daftar harga asli produk bersangkutan, dan pada jumlah_terjual dengan nilai median.

5. Menyeragamkan kolom tanggal ke format datetime standar.

Rencana Manipulasi
Data

1. Filter: Mengabaikan/memfilter data transaksi bernilai salah/outlier (seperti jumlah_terjual == 500). |
2. Sort: Mengurutkan transaksi berdasarkan total_penjualan secara descending. |
3. Kolom turunan: Membuat kolom total_penjualan (jumlah_terjual $\times$ harga_satuan). |
4. Groupby/agregasi: Mengelompokkan data berdasarkan kategori dan nama_produk untuk menghitung total pendapatan dan total unit terjual.

Jadwal Kerja

1. P3 (Loading & Inspection): Hari ke 1 |
2. P4 (Cleaning): Hari ke 2 |
3. P5 (Manipulation): Hari ke 3 |
4. P6 (Uji & Presentasi): Hari ke 4